scapy/contrib/iec104.py

```diff
@@ -0,0 +1,596 @@
+    IEC104_COT_INRO3: 'queried by query to group 3 (inro3)',
+    IEC104_COT_INRO4: 'queried by query to group 4 (inro4)',
+    IEC104_COT_INRO5: 'queried by query to group 5 (inro5)',
+    IEC104_COT_INRO6: 'queried by query to group 6 (inro6)',
+    IEC104_COT_INRO7: 'queried by query to group 7 (inro7)',
+    IEC104_COT_INRO8: 'queried by query to group 8 (inro8)',
+    IEC104_COT_INRO9: 'queried by query to group 9 (inro9)',
+    IEC104_COT_INRO10: 'queried by query to group 10 (inro10)',
+    IEC104_COT_INRO11: 'queried by query to group 11 (inro11)',
+    IEC104_COT_INRO12: 'queried by query to group 12 (inro12)',
+    IEC104_COT_INRO13: 'queried by query to group 13 (inro13)',
+    IEC104_COT_INRO14: 'queried by query to group 14 (inro14)',
+    IEC104_COT_INRO15: 'queried by query to group 15 (inro15)',
+    IEC104_COT_INRO16: 'queried by query to group 16 (inro16)',
+    IEC104_COT_REQCOGEN: 'queried by counter general interrogation (reqcogen)',
+    IEC104_COT_REQCO1: 'queried by query to counter group 1 (reqco1)',
+    IEC104_COT_REQCO2: 'queried by query to counter group 2 (reqco2)',
+    IEC104_COT_REQCO3: 'queried by query to counter group 3 (reqco3)',
+    IEC104_COT_REQCO4: 'queried by query to counter group 4 (reqco4)',
+    IEC104_COT_RESERVED_42: 'reserved_42',
+    IEC104_COT_RESERVED_43: 'reserved_43',
+    IEC104_COT_UNKNOWN_TYPE_CODE: 'unknown type code',
+    IEC104_COT_UNKNOWN_TRANSMIT_REASON: 'unknown transmit reason',
+    IEC104_COT_UNKNOWN_COMMON_ADDRESS_OF_ASDU: 'unknown common address of ASDU',
+    IEC104_COT_UNKNOWN_ADDRESS_OF_INFORMATION_OBJECT: 'unknown address of information object',
+    IEC104_COT_PRIVATE_48: 'private_48',
+    IEC104_COT_PRIVATE_49: 'private_49',
+    IEC104_COT_PRIVATE_50: 'private_50',
+    IEC104_COT_PRIVATE_51: 'private_51',
+    IEC104_COT_PRIVATE_52: 'private_52',
+    IEC104_COT_PRIVATE_53: 'private_53',
+    IEC104_COT_PRIVATE_54: 'private_54',
+    IEC104_COT_PRIVATE_55: 'private_55',
+    IEC104_COT_PRIVATE_56: 'private_56',
+    IEC104_COT_PRIVATE_57: 'private_57',
+    IEC104_COT_PRIVATE_58: 'private_58',
+    IEC104_COT_PRIVATE_59: 'private_59',
+    IEC104_COT_PRIVATE_60: 'private_60',
+    IEC104_COT_PRIVATE_61: 'private_61',
+    IEC104_COT_PRIVATE_62: 'private_62',
+    IEC104_COT_PRIVATE_63: 'private_63'
+}
+
+IEC104_APDU_TYPE_UNKNOWN = 0x00
+IEC104_APDU_TYPE_I_SEQ_IOA = 0x01
+IEC104_APDU_TYPE_I_SINGLE_IOA = 0x02
+IEC104_APDU_TYPE_U = 0x03
+IEC104_APDU_TYPE_S = 0x04
+
+
+def _iec104_apci_type_from_packet(data):
+    """
+    the type of the message is encoded in octet 1..4
+
+                 oct 1, bit 1   2       oct 3, bit 1
+    I Message               0  1|0                 0
+    S Message               1   0                  0
+    U Message               1   1                  0
+
+
+    see EN 60870-5-104:2006, sec. 5 (p. 13, fig. 6,7,8)
+    """
+
+    oct_1 = orb(data[2])
+    oct_3 = orb(data[4])
+
+    oct_1_bit_1 = True if oct_1 & 1 else False
```

Comment:
Use `bool(oct_1 & 1)` same below

Reply:
done here: d007e72

and believe me or not: the bit numbering comes straight from the standard.